xaitk_saliency/interfaces/vis_sal_image_classifier_blackbox.py

```diff
@@ -0,0 +1,99 @@
+    matrices, one for each class output by the classifier blackbox.
+
+    An input classifier blackbox needs to be input, which requires some
+    specification in how to operate the blackbox.
+    The `smqtk_classifier.ClassifyImage` abstract interface is utilized for
+    this purpose in order to provide some minimal form that a blackbox
+    classifier must have: be able to classify an image into confidences for
+    some number of class labels.
+
+    Generated visual saliency heat-map for each input class as a
+    float-type `numpy.ndarray` of shape `[nClasses x H x W]`.
+    """
+
+    def generate(
+        self,
+        ref_image: np.ndarray,
+        blackbox: ClassifyImage
+    ) -> np.ndarray:
+        """
+        Generate per-class visual saliency heatmaps for some classifier
+        blackbox over some image of interest.
+
+        The input reference image is expected to be in matrix form and be in
+        either a `H x W` or `H x W x C` shape format.
+
+        Output saliency map matrix should be in the shape `nClasses x H x W`,
+        be floating-point typed and be composed of values in the `[-1, 1]`
+        range.
+        `nClasses` should be the quantity of unique class labels output by the
+        given classifier blackbox.
+        While specific algorithms determine the quantity of heatmaps returned
+        the height and width of returned heatmaps should be consistent with the
+        input image, i.e. the `H` and `W` dimensions should match in size to
+        the reference image's `H` and `W` dimensions.
+        Positive values of the saliency heat-maps indicate regions which
+        increase respective class confidence scores, while negative values
+        indicate regions which decrease respective class confidence scores
+        according to the given blackbox classifier.
+
+        :param ref_image: Reference image over which visual saliency heatmaps
+            will be generated.
+        :param blackbox: The blackbox classifier handle to perform arbitrary
+            operations on in order to deduce visual saliency.
+
+        :raises ShapeMismatchError: The implementation result visual saliency
+            heatmap matrix did not have matching height and width components to
+            the reference image.
+
+        :return: A number of visual saliency heatmaps equivalent in number to
+            the quantity of class labels output by the blackbox classifier.
+        """
+        # This is a template method that enforces standard input and result
+        # checks.
+        if ref_image.ndim not in (2, 3):
+            raise ValueError(f"Input image matrix has an unexpected number of dimensions: {ref_image.ndim}")
+        output = self._generate(ref_image, blackbox)
+        # Check that the saliency heatmaps' shape matches the reference image.
+        if output.shape[1:] != ref_image.shape[:2]:
+            raise ShapeMismatchError(
```

Comment:
Is this a newly added constraint for the high-level interface (and if so, why)? Using the low-level interface, I think it is possible/allowed for the reference image and the generated saliency maps to be different shapes (you mentioned not leaking image size to the blackbox). Should we make this a constraint throughout our API?

Reply:
I didn't think this is a new constraint. We previously documented a constraint on `PerturbImage` that output masks are of equivalent HxW to the input reference image. Masks are then input to the generator (`ImageClassifierSaliencyMapGenerator` in this case) that is documented to output masks with equivalent HxW of the masks input. So, by transitivity, the output heatmap is to have the same HxW of the input image.

Reply:
Yes, you're right- somehow I got confused since I remembered in our example notebooks input images were 480x640, then resized to 224x224 before passing into the black box classifier. I thought the output saliency maps were 224x224, but it looks like they are the same size as the reference input image.